Deployment modes collector (#294)
elastic/opentelemetry: docs/reference/edot-collector/modes.md

## EDOT Collector as Agent

In Agent mode, the EDOT Collector runs close to the data source, collecting telemetry data directly from the local environment. A Collector in Agent mode usually runs on the same host or virtual machine as the application or infrastructure component it is monitoring, or as a sidecar container or daemonset in Kubernetes.

The Collector in Agent mode can also perform initial processing, filtering, and batching of data before sending the data to a Collector running in Gateway mode or directly to a Managed OTLP Endpoint.

Use the EDOT Collector in Agent mode when:

- You need to collect data directly from hosts or applications.
- You want to minimize network traffic by performing initial filtering at the source.
- You have a simple deployment with a small number of hosts.

## EDOT Collector as Gateway

In Gateway mode, the EDOT Collector acts as a central aggregation point, receiving data from multiple Agent collectors or instrumented applications before forwarding it to Elastic. The flexibility of the Gateway mode allows it to perform different operations at different stages of the data pipeline.

Use the EDOT Collector in Gateway mode when:

> - You have multiple data sources or agents that need centralized processing.

Reviewer: ```suggestion
- You have multiple data sources or agents that need centralized processing or enrichment.
```

- You need to implement organization-wide processing rules.
- You want to reduce the number of connections to your Elastic backend.
- You need advanced pre-processing before data reaches Elastic.
- You're using a self-managed Elasticsearch deployment (required for APM functionality).

### Gateway requirements for self-managed environments

> For self-managed Elastic environments, you need a Gateway Collector configured with the Elasticsearch exporter. This is because the Gateway includes essential components for APM functionality:

Reviewer: ```suggestion
For self-managed Elastic environments, you need a Gateway Collector deployed alongside your Elastic Stack. The EDOT Collector in gateway mode exposes a scalable OTLP endpoint and performs data processing required for APM functionality.

This is the only case where using the Elasticsearch exporter is recommended. In all other EDOT Collector deployments described in this guide, we recommend using the OTLP exporter.
```

> 

Reviewer: ```suggestion

#### Data processing required for APM functionality in self-managed Elastic
```

- The `elastictrace` processor enriches trace data with additional attributes that improve the user experience in Elastic Observability UIs.
- The `elasticapm` connector generates pre-aggregated APM metrics from tracing data.

> The Gateway also handles the translation between OpenTelemetry data formats and Elastics-compatible formats when needed, and uses the routing connector to direct different types of telemetry data to the appropriate pipelines.

Reviewer: This reads as schema translation, we want to stay away from even using the word translation as this confuses most people. I'd rather avoid going into the level of detail as most people won't care.

Reviewer: ```suggestion
In this case EDOT Collector as a gateway also handles the routing the different types of telemetry data to the relevant indices.
```

> ### Gateway pattern beyond self-managed deployments

Reviewer: I think we can address most of the content of this section at the top where we define the gateway mode

---
navigation_title: Deployment modes
description: Deployment modes for the Elastic Distribution of OpenTelemetry (EDOT) Collector, including Agent and Gateway modes and when to use each.
applies_to:
  stack:
  serverless:
    observability:
products:
  - id: cloud-serverless
  - id: observability
  - id: edot-collector
---

# EDOT Collector deployment modes

> You can deploy EDOT Collector in different modes to meet your architectural requirements. The two primary deployment modes are Agent and Gateway. Depending on whether you're using Elastic self-managed, {{ecloud}}, or {{serverless-full}}, various EDOT Collector instances, deployed in each mode may be required to support the target architecture.

Reviewer 2: ```suggestion
You can deploy the EDOT Collector in different modes to meet your architectural needs. The two primary deployment modes are Agent and Gateway. Depending on your deployment type — Elastic self-managed, {{ecloud}}, or {{serverless-full}} — various EDOT Collector instances may be required in each mode to support the target architecture.
```

Use the information in the following sections to better understand deployment modes and patterns for your specific environment.

## EDOT Collector as Agent

In Agent mode, the EDOT Collector runs close to the data source, collecting telemetry data directly from the local environment. A Collector in Agent mode usually runs on the same host or virtual machine as the application or infrastructure component it is monitoring, or as a sidecar container or daemonset in Kubernetes.

Use the EDOT Collector in Agent mode when:

- You need to collect data directly from hosts or applications.

> - You have a simple deployment with a small number of hosts.

Reviewer 2: What does “simple” stand for?

Author: Weasel word detected! Let me remove it.

## EDOT Collector as Gateway

> In Gateway mode, the EDOT Collector acts as a central aggregation point, receiving data from multiple EDOT Agent collectors or instrumented applications before forwarding it to Elastic. The flexibility of the Gateway mode allows it to centralize scaling needs and data transformation operations of the data pipeline.

Reviewer 2: ```suggestion
In Gateway mode, the EDOT Collector acts as a central aggregation point, receiving data from multiple EDOT Agent collectors or instrumented applications before forwarding it to Elastic. The flexibility of the Gateway mode allows the Collector to centralize the scaling needs and data transformation operations of the data pipeline.
```

Use the EDOT Collector in Gateway mode when:

- You have multiple data sources or agents that need centralized processing or enrichment.
- You need to implement organization-wide processing rules.
- You want to reduce the number of connections to your Elastic backend.
- You need advanced pre-processing before data reaches Elastic.

> - You're using a self-managed Elasticsearch deployment (required for APM functionality).

Reviewer 2: ```suggestion
- You're using a self-managed {{es}} deployment (required for APM functionality).
```

> - You want to filter telemetry before it is shipped over the network to Elastic

Reviewer 2: ```suggestion
- You want to filter telemetry before it is shipped over the network to Elastic.
```

The Gateway pattern isn't exclusive to self-managed Elastic deployments. It's a general OpenTelemetry pattern that provides benefits in various scenarios:

- Kubernetes deployments: A Gateway collector centralizes cluster-level telemetry from multiple node-level Agent collectors.

> - Multi-Region deployments: Regional Gateway collectors aggregate data from multiple Agents before sending to a central destination.

Reviewer 2: ```suggestion
- Multi-region deployments: Regional Gateway collectors aggregate data from multiple Agents before sending to a central destination.
```

> - High-Volume Environments: Gateway collectors provide buffering and batching to handle high volumes of telemetry data.

Reviewer 2: ```suggestion
- High-volume environments: Gateway collectors provide buffering and batching to handle high volumes of telemetry data.
```

> - Complex Processing: When advanced data transformation or filtering is needed before data reaches its destination.

Reviewer 2: ```suggestion
- Complex processing: When advanced data transformation or filtering is needed before data reaches its destination.
```

### Gateway requirements for self-managed environments

> For self-managed Elastic environments, you need a Gateway Collector deployed alongside your Elastic Stack. The EDOT Collector in Gateway mode exposes a scalable OTLP endpoint and performs data processing required for APM functionality.

Reviewer 2: ```suggestion
For self-managed Elastic environments, you need a Gateway Collector deployed alongside your {{sack}}. The EDOT Collector in Gateway mode exposes a scalable OTLP endpoint, and performs data processing required for APM functionality.
```

> This is the only case where using the Elasticsearch exporter is recommended. In all other EDOT Collector deployments described in this guide, use the OTLP exporter.

Reviewer 2: Maybe this sentence should be highlighted, and placed in a note? Otherwise someone not using a self-managed environment may skip this section and miss this important information?

```suggestion
This is the only case where using the {{es}} exporter is recommended. In all other EDOT Collector deployments described in this guide, use the OTLP exporter.
```

#### Required components for APM functionality in self-managed Elastic

The following components are required for APM functionality in self-managed Elastic:

- The `elastictrace` processor enriches trace data with additional attributes that improve the user experience in Elastic Observability UIs.

> - The `elasticapm` connector generates pre-aggregated APM metrics from tracing data.

Reviewer 2: Not sure about this, but shouldn’t this be an adjective for data? (as opposed to making this look like a verbal phrase)

```suggestion
- The `elasticapm` connector generates pre-aggregated APM metrics from trace data.
```

> In this case EDOT Collector as a Gateway also handles the routing the different types of telemetry data to the relevant indices.

Reviewer 2: ```suggestion
In this case, EDOT Collector as a Gateway also handles routing of the different types of telemetry data to the relevant indices.
```

## Deployment in Kubernetes environments

> In Kubernetes environments, EDOT Collectors are typically deployed in three distinct modes that work together to provide comprehensive observability:

Reviewer 2: Not sure I’d end this sentence with a `:` if I’m not listing the options right below it. If you can’t mention the three options here, maybe just use a period.

Author: +1 (I moved content around)

### Agent mode in Kubernetes

In Kubernetes, the Agent mode is implemented in two forms:

| Mode | Deployment | Functions |
|------|------------|-----------|

> | Daemon mode | DaemonSet on every node. | - Collects node-local logs and host metrics.<br>- Receives telemetry from applications instrumented with OpenTelemetry SDKs running on the node.<br>- Enriches application telemetry with resource information such as host and Kubernetes metadata.<br>- Forwards all data to the Gateway Collector using the OTLP protocol. |

Reviewer 2: ```suggestion
| Daemon mode | DaemonSet on every node | - Collects node-local logs and host metrics.<br>- Receives telemetry data from applications instrumented with OpenTelemetry SDKs running on the node.<br>- Enriches application telemetry data with resource information such as host and Kubernetes metadata.<br>- Forwards all data to the Gateway Collector using the OTLP protocol. |
```

| Cluster mode | Centralized service. | - Collects Kubernetes cluster-level metrics from the Kubernetes API.<br>- Monitors cluster-wide resources that aren't specific to individual nodes.<br>- Forwards collected data to the Gateway Collector using the OTLP protocol. |

### Gateway mode in Kubernetes

The Gateway Collector in Kubernetes receives data from all Daemon and Cluster collectors. The Gateway performs additional pre-processing and aggregation for self-managed and Elastic Cloud Hosted deployments, and handles the final export to the appropriate Elastic backend.

This multi-tier architecture in Kubernetes provides an efficient way to collect and process telemetry data at different levels of the infrastructure while minimizing resource usage and network traffic.

For more details on Kubernetes deployment architecture, see [Kubernetes environments](../architecture/k8s.md).

## Direct EDOT SDK to Managed OTLP (No Collector)

> In some scenarios, you don't need a Collector at all. EDOT SDKs can send telemetry data directly to Elastic's Managed OTLP Endpoint. This is the simplest deployment pattern for getting application telemetry into Elastic Observability.

Reviewer 2: ```suggestion
In some scenarios, you don't need a Collector at all. EDOT SDKs can send telemetry data directly to Elastic's Managed OTLP Endpoint. This is the simplest deployment pattern for getting application telemetry data into Elastic Observability.
```
